src/display/node_stream.js

```diff
@@ -0,0 +1,351 @@
+  }
+
+  _setReadableStream(readableStream) {
+    this._readableStream = readableStream;
+    readableStream.on('readable', () => {
+      this._readCapability.resolve();
+    });
+
+    readableStream.on('end', () => {
+      // Destroy readable to minimize resource usage.
+      readableStream.destroy();
+      this._done = true;
+      this._readCapability.resolve();
+    });
+
+    readableStream.on('error', (reason) => {
+      this._errored = true;
+      this._reason = reason;
+      this._readCapability.resolve();
+    });
+  }
+}
+
+class BaseRangeReader {
+  constructor(stream) {
+    this._url = stream.url;
+    this._done = false;
+    this._errored = false;
+    this._reason = null;
+    this.onProgress = null;
+    this._loaded = 0;
+    this._readableStream = null;
+    this._readCapability = createPromiseCapability();
+
+    this._isStreamingSupported = !stream.source.disableStream;
+  }
+
+  get isStreamingSupported() {
+    return this._isStreamingSupported;
+  }
+
+  read() {
+    return this._readCapability.promise.then(() => {
+      if (this._done) {
+        return Promise.resolve({ value: undefined, done: true, });
+      }
+      if (this._errored) {
+        return Promise.reject(this._reason);
+      }
+
+      let chunk = this._readableStream.read();
+      if (chunk === null) {
+        this._readCapability = createPromiseCapability();
+        return this.read();
+      }
+      this._loaded += chunk.length;
+      if (this.onProgress) {
+        this.onProgress({ loaded: this._loaded, });
+      }
+      return Promise.resolve({ value: chunk, done: false, });
+    });
+  }
+
+  cancel(reason) {
+    this._readableStream.destroy(reason);
+  }
+
+  _setReadableStream(readableStream) {
+    this._readableStream = readableStream;
+    readableStream.on('readable', () => {
+      this._readCapability.resolve();
+    });
+
+    readableStream.on('end', () => {
+      readableStream.destroy();
+      this._done = true;
+      this._readCapability.resolve();
+    });
+
+    readableStream.on('error', (reason) => {
+      this._errored = true;
+      this._reason = reason;
+      this._readCapability.resolve();
+    });
+  }
+}
+
+function createRequestOptions(url, headers) {
+  let agentOptions = {
+    keepAlive: true,
+    maxSockets: 10,
+  };
+  let agent = url.protocol === 'http:' ? new http.Agent(agentOptions) :
+    new https.Agent(agentOptions);
+
+  return {
+    protocol: url.protocol,
+    auth: url.auth,
+    host: url.host,
+    port: url.port,
+    path: url.path,
+    method: 'GET',
+    headers,
+    agent,
+  };
+}
+
+class PDFNodeStreamFullReader extends BaseFullReader {
+  constructor(stream) {
+    super(stream);
+
+    let handleResponse = (response) => {
+      this._headersCapability.resolve();
+      this._setReadableStream(response);
+
+      let { allowRangeRequests, suggestedLength, } =
+      validateRangeRequestCapabilities({
+        getResponseHeader: this.getResponseHeader.bind(this),
```

Comment:
remove getResponseHeader function below and inline the response.headers access here